[GTK4] WebKitWebView::context-menu errors out on GTK4

alice-mkh opened the pull request:
#### 7f793269519256ea5f2b8b420ed02578d4ce6f2c
<pre>
[GTK4] WebKitWebView::context-menu errors out on GTK4
<a href="https://bugs.webkit.org/show_bug.cgi?id=233785">https://bugs.webkit.org/show_bug.cgi?id=233785</a>

Reviewed by Michael Catanzaro.

GdkEvent is not a boxed type or an object, and so needs a custom marshaler.

Fix a separate crash in WebContextMenuProxyGtk, it had never been noticed
as the signal crash happened before that.

Add GdkEvent support to GRefPtr on GTK4.

* Source/WebCore/platform/gtk/GRefPtrGtk.cpp:
(WTF::refGPtr):
(WTF::derefGPtr):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(gValueGetEvent):
(webkitWebViewContextMenuMarshal):
(webkitWebViewContextMenuMarshalVa):
(webkit_web_view_class_init):
* Source/WebKit/UIProcess/gtk/WebContextMenuProxyGtk.cpp:
(WebKit::menuWidgetHasItems):

Canonical link: <a href="https://commits.webkit.org/252908@main">https://commits.webkit.org/252908@main</a>
</pre>

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+
+static void webkitWebViewContextMenuMarshal(GClosure* closure, GValue* returnValue, guint nParams, const GValue* params, gpointer, gpointer marshalData)
+{
+    GCClosure* cc = (GCClosure *) closure;
```
```suggestion
    auto* cc = static_cast<GCClosure*>(closure);
```

No C-style casts unless C++ casts are impossible (very rare). Also, no point in writing out the name of the type twice. Also, your * must hang on the type.

alice-mkh replied:
Ah yeah. This code was based on glib-genmarshal output so the starting point was C. :)

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+    auto* event = gValueGetEvent(&params[2]);
+    auto* result = WEBKIT_HIT_TEST_RESULT(g_value_get_object(&params[3]));
+
+    ContextMenuCallback callback = (ContextMenuCallback) (marshalData ? marshalData : cc->callback);
```
```suggestion
    auto callback = static_cast<ContextMenuCallback>(marshalData ? marshalData : cc->callback);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+    ContextMenuCallback callback = (ContextMenuCallback) (marshalData ? marshalData : cc->callback);
+    gboolean ret = callback(data1, menu, event, result, data2);
+
+    g_value_set_boolean(returnValue, ret);
```
Style nit: we generally use fewer blank lines:

```suggestion
    gboolean ret = callback(data1, menu, event, result, data2);
    g_value_set_boolean(returnValue, ret);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+
+static void webkitWebViewContextMenuMarshalVa(GClosure *closure, GValue *returnValue, gpointer instance, va_list args, gpointer marshalData, int, GType*)
+{
+    GCClosure* cc = (GCClosure *) closure;
```
```suggestion
    auto* cc = static_cast<GCClosure*>(closure);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+    g_value_set_boolean(returnValue, ret);
+}
+
+static void webkitWebViewContextMenuMarshalVa(GClosure *closure, GValue *returnValue, gpointer instance, va_list args, gpointer marshalData, int, GType*)
```
```suggestion
static void webkitWebViewContextMenuMarshalVa(GClosure* closure, GValue* returnValue, gpointer instance, va_list args, gpointer marshalData, int, GType*)
```

alice-mkh replied:
Hm, surprised the style checker didn't find this one.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+
+    va_end(argsCopy);
+
+    ContextMenuCallback callback = (ContextMenuCallback) (marshalData ? marshalData : cc->callback);
```
```suggestion
    auto callback = static_cast<ContextMenuCallback>(marshalData ? marshalData : cc->callback);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+
+    auto* menu = WEBKIT_CONTEXT_MENU(g_object_ref(va_arg(argsCopy, gpointer)));
+    auto* event = gdk_event_ref(reinterpret_cast<GdkEvent*>(va_arg(argsCopy, gpointer)));
+    auto* result = WEBKIT_HIT_TEST_RESULT(g_object_ref(va_arg(argsCopy, gpointer)));
```
Ownership via raw pointers is prohibited. Need to use the smart pointers, e.g.:

```suggestion
    GRefPtr<WebKitContextMenu> menu = WEBKIT_CONTEXT_MENU(va_arg(argsCopy, gpointer));
    GRefPtr<GdkEvent> event = reinterpret_cast<GdkEvent*>(va_arg(argsCopy, gpointer));
    GRefPtr<WebKitHitTestResult> result = WEBKIT_HIT_TEST_RESULT(va_arg(argsCopy, gpointer));
```

alice-mkh replied:
GRefPtr doesn't work with GdkEvent atm.

mcatanzaro replied:
You get to fix that then. Should be very easy. Just copy/paste in GRefPtrGtk.h and GRefPtrGtk.cpp and have it use gdk_event_ref/unref.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1048,6 +1048,77 @@ static gboolean webkitWebViewAccumulatorObjectHandled(GSignalInvocationHint*, GV
+
+    g_object_unref(menu);
+    gdk_event_unref(event);
+    g_object_unref(result);
```
```suggestion
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2133,15 +2204,27 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
         WEBKIT_TYPE_HIT_TEST_RESULT);
+#if USE(GTK4)
+    g_signal_set_va_marshaller(signals[CONTEXT_MENU],    G_TYPE_FROM_CLASS(webViewClass),
+        webkitWebViewContextMenuMarshalVa);
```
Style nit:

```suggestion
    g_signal_set_va_marshaller(signals[CONTEXT_MENU], G_TYPE_FROM_CLASS(webViewClass), webkitWebViewContextMenuMarshalVa);
```

mcatanzaro reviewed (changes requested):
What's the problematic difference between GTK 3 vs. GTK 4 here? The problem that prevents this marshaller from being used for GTK 3?

alice-mkh committed:
[GTK4] WebKitWebView::context-menu errors out on GTK4
https://bugs.webkit.org/show_bug.cgi?id=233785

Reviewed by Michael Catanzaro.

GdkEvent is not a boxed type or an object, and so needs a custom marshaler.

Fix a separate crash in WebContextMenuProxyGtk, it had never been noticed
as the signal crash happened before that.

Add GdkEvent support to GRefPtr on GTK4.

* Source/WebCore/platform/gtk/GRefPtrGtk.cpp:
(WTF::refGPtr):
(WTF::derefGPtr):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(gValueGetEvent):
(webkitWebViewContextMenuMarshal):
(webkitWebViewContextMenuMarshalVa):
(webkit_web_view_class_init):
* Source/WebKit/UIProcess/gtk/WebContextMenuProxyGtk.cpp:
(WebKit::menuWidgetHasItems):

Canonical link: https://commits.webkit.org/252908@main
